test/e2e/storage/testsuites/base.go

```diff
@@ -0,0 +1,319 @@
+	GetTestPatterns() []testpatterns.TestPattern
+	// SkipUnsupportedTest skips the test if this TestSuite is not suitable to be tested with the combination of TestPattern and TestDriver
+	SkipUnsupportedTest(testpatterns.TestPattern, drivers.TestDriver)
+	// InitTestResource initializes and returns the TestResource to be used with this TestSuite
+	InitTestResource() TestResource
+	// CreateTestMetadata creates and returns TestInput that is only used for print context string in parse phase of test loop
+	CreateTestMetadata(TestSuite, testpatterns.TestPattern) TestInput
+	// CreateTestInput creates and returns TestInput that is passed to test function
+	CreateTestInput(testpatterns.TestPattern, TestResource) TestInput
+	// GetTestFunc returns a test function to be tested for the combination of TestPattern and TestDriver
+	GetTestFunc(testpatterns.TestPattern, drivers.TestDriver) func(*TestInput)
+}
+
+// TestResource represents an interface for resources that is used by TestSuite
+type TestResource interface {
+	// SetupResource sets up test resources to be used for the tests with the
+	// combination of TestDriver and TestPattern
+	SetupResource(drivers.TestDriver, testpatterns.TestPattern)
+	// CleanupResource clean up the test resources created in SetupResource
+	CleanupResource()
+}
+
+// TestInput represents an interface for an input to be passed for the test function
+type TestInput interface {
+	// GetTestMetadataString returns a string that represents test metadata
+	GetTestMetadataString() string
+}
+
+// testMetadata is an implementation of TestInput interface
+// This struct should be included in all the implementation of TestInput interface.
+type testMetadata struct {
+	testSuite   TestSuite
+	testPattern testpatterns.TestPattern
+}
+
+// GetTestMetadataString returns a string that represents test metadata
+// This function is used in parse stage of the test loop and should not be used in execution stage.
+func (t testMetadata) GetTestMetadataString() string {
+	return fmt.Sprintf("[TestSuite: %s][TestPattern: %s]", t.testSuite.GetName(), t.testPattern.Name)
+}
+
+// InitTestMetadata returns testMetadata
+// This function is used in parse stage of the test loop and should not be used in execution stage.
+func InitTestMetadata(ts TestSuite, tp testpatterns.TestPattern) testMetadata {
+	return testMetadata{
+		testSuite:   ts,
+		testPattern: tp,
+	}
+}
+
+// generateTestFunc generates a function to be passed to GetTestFunc for each TestSuites.
+// This function takes testFunc as a interface{} and test input as a pointer for TestInput interface.
+// Type assertion for testFunc is required in parse time of ginkgo and
+// that for testInput is required in BeforeEach, or test execution time of ginkgo.
+func generateTestFunc(testFunc interface{}) func(*TestInput) {
+	return func(ti *TestInput) {
+		Context((*ti).GetTestMetadataString(), func() {
+			var ok bool
+
+			switch fn := testFunc.(type) {
+			case func(*volumesTestInput):
```

Comment:
This is leaking test suite implementation details into the framework.  Is there a way we can avoid having a TestInput at all?

For example, could `testSubpath` include the Context() and BeforeEach() that it needs to setup all the necessary inputs (including the genericTestResource) ?  Then in the framework, you don't need a CreateTestInput method.

Reply:
Could you explain a little bit more about your idea?
 
I still don't understand your idea completely, but I'm worry about that it would require all TestSuite implement the similar codes. Also, to leave testSubPath() accept testInput will allow other codes reuse testSubPath(), which I guess would still benefit some of the existing tests.
(I agree with you that current implementation is not optimal, though.)

Reply:
Something like:

```
func testSubpath(driver, testPattern) {
  Context("Subpath") , func() {
     BeforeEach(func() {
          // Setup all the test inputs and resources here
    })

     After Each(func() {
         // Cleanup
    }

   It("test case 1")

   It("test case 2")

  }
}